### ISSUE TITLE
Bug: Could not find a declaration file for module 'utils'

### ISSUE THREAD

**Issue by LeeMoonki** [typescript-eslint/typescript-eslint#7762]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
import { ESLintUtils, TSESTree } from '@typescript-eslint/utils';
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",
    "outDir": "lib",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "strictNullChecks": true,
    "skipLibCheck": true
  },
  "include": ["src", "index.ts"],
  "watchOptions": {
    "excludeDirectories": ["**/node_modules", "lib", "tests", "tests-etc"]
  }
}
```

### Expected Result

I can use the module of `@typescript-eslint/utils`.

### Actual Result

The following error occurs.
```
Cannot find module @typescript-eslint/utils or its corresponding type declarations
```

### Additional Info

When I use [@typescript-eslint/utils](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/utils), the error 
```
Cannot find module @typescript-eslint/utils or its corresponding type declarations
```
occurred.

The project structure is
```
.
├── packages/
│   └── eslint-plugin-foo/
│       ├── node_modules/
│       │   └── @typescript-eslint/
│       │       └── utils
│       ├── src/
│       │   └── rules/
│       │       └── some-rule.ts
│       └── tsconfig.json
└── tsconfig.json
```

When I use the package like,
```typescript
import { ESLintUtils, TSESTree } from '@typescript-eslint/utils/dist';
...
```
the error disappear. But when I run a test for it, the following error shown
```
Package subpath './dist' is not defined by "exports" in somepath/packages/eslint-plugin-foo/node_modules/@typescript-eslint/utils/package.json
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/utils`             | `6.8.0` |

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/7284
